lib/elixir/lib/module/types.ex

```diff
@@ -0,0 +1,850 @@
+  end
+
+  defp new_var(var, context) do
+    case :maps.find(var_name(var), context.vars) do
+      {:ok, type} ->
+        {type, context}
+
+      :error ->
+        type = {:var, context.counter}
+        vars = :maps.put(var_name(var), type, context.vars)
+        types_to_vars = :maps.put(context.counter, var, context.types_to_vars)
+        types = :maps.put(context.counter, :unbound, context.types)
+        traces = :maps.put(context.counter, [], context.traces)
+        counter = context.counter + 1
+
+        context = %{
+          context
+          | vars: vars,
+            types_to_vars: types_to_vars,
+            types: types,
+            traces: traces,
+            counter: counter
+        }
+
+        {type, context}
+    end
+  end
+
+  defp refine_var(var, type, context) do
+    types = :maps.put(var, type, context.types)
+    trace_var(var, type, %{context | types: types})
+  end
+
+  defp trace_var(var, type, context) do
+    line = get_meta(hd(context.expr_stack))[:line]
+    trace = {type, context.expr_stack, {context.file, line}}
+    traces = :maps.update_with(var, &[trace | &1], context.traces)
+    %{context | traces: traces}
+  end
+
+  defp var_name({name, _meta, context}), do: {name, context}
+
+  # Check if a variable is recursive and incompatible with itself
+  # Bad: `{var} = var`
+  # Good: `x = y; y = z; z = x`
+  defp recursive_type?({:var, var} = parent, parents, context) do
+    case :maps.get(var, context.types) do
+      :unbound ->
+        false
+
+      type ->
+        if type in parents do
+          not Enum.all?(parents, &match?({:var, _}, &1))
+        else
+          recursive_type?(type, [parent | parents], context)
+        end
+    end
+  end
+
+  defp recursive_type?({:cons, left, right} = parent, parents, context) do
+    recursive_type?(left, [parent | parents], context) or
+      recursive_type?(right, [parent | parents], context)
+  end
+
+  defp recursive_type?({:tuple, types} = parent, parents, context) do
+    Enum.any?(types, &recursive_type?(&1, [parent | parents], context))
+  end
+
+  defp recursive_type?({:map, pairs} = parent, parents, context) do
+    Enum.any?(pairs, fn {key, value} ->
+      recursive_type?(key, [parent | parents], context) or
+        recursive_type?(value, [parent | parents], context)
+    end)
+  end
+
+  defp recursive_type?(_other, _parents, _context) do
+    false
+  end
+
+  # Collect binary type specifiers,
+  # from `<<pattern::integer-size(10)>>` collect `integer`
+  defp collect_specifiers({:-, _meta, [left, right]}) do
+    [collect_specifiers(left), collect_specifiers(right)]
+  end
+
+  defp collect_specifiers({specifier, _meta, []}) do
+    [specifier]
+  end
+
+  defp collect_specifiers({name, _meta, context}) when is_atom(name) and is_atom(context) do
+    [name]
+  end
+
+  defp collect_specifiers(_other) do
+    []
+  end
+
+  defp subtype?({:literal, boolean}, :boolean, _context) when is_boolean(boolean), do: true
```

Comment:
We should start moving those functions that work on types to  a separate module so we can extensively unit test them.